### ISSUE TITLE
Pattern matching based on generic type

### ISSUE THREAD

**Issue by colin-jack** [dotnet/roslyn#178]
One use for pattern matching could be to choose appropriate code path based on a generic type. Say we want to map objects of types Foo and Bar to a new object of type Target. However the code that needs to do the mapping doesn't know if we have a Foo, a Bar, or something else:

```
        public class Target { ... }

        public class Foo { ... }

        public class Bar { ... }

        public class FooMapper {
            public Target Map(Foo mapFrom) {...}
        }

        public class BarMapper {
            public Target Map(Bar mapFrom) {...}
        }
        
        private Target Map<TFrom>(TFrom mappingFrom) = TFrom match { 
            case Foo => new FooMapper().Map(mappingFrom) // NOTE: No need to cast mappingFrom, we only match the case if its a Foo so no need to do "FooMapper().Map((Foo)mappingFrom)"
            case Bar => new BarMapper().Map(mappingFrom)
            default => throw new InvalidOprationException(..); // NOTE: or maybe no match implicitly produces exception
        }
```

There are alternatives already but I think this Scala style matching shows whats happening quite cleanly. Obviously you could add type constraints too:

```
    private Target Map<TFrom>(TFrom mappingFrom) where TFrom: IModel = TFrom match { 

```

**Comment by Pilchie** [comment]
If you look at the notes in Issue #98, it already discusses a proposal for pattern matching, including support for algebraic data types.  In fact, there is already a prototype of some of it.  @gafter Will likely be putting up a formal proposal issue for it soon.

**Comment by gafter** [comment]
@Pilchie The switch expression in this proposal is switching on a *generic type parameter*, not an *expression*. What is proposed is not covered by the existing proposal.

**Comment by gafter** [comment]
@colin-jack Does #206 cover the use cases this is intended to address?

**Comment by colin-jack** [comment]
@gafter 
Ta for replying. I think I did look at #206 at time but didn't find it overly accessible, I probably need to sit down and work through it so have closed this issue.

**Comment by alrz** [comment]
@colin-jack 

>
        private Target Map<TFrom>(TFrom mappingFrom) = TFrom match { 
            case Foo => new FooMapper().Map(mappingFrom) // NOTE: No need to cast mappingFrom, we only match the case if its a Foo so no need to do "FooMapper().Map((Foo)mappingFrom)"
            case Bar => new BarMapper().Map(mappingFrom)
            default => throw new InvalidOprationException(..); // NOTE: or maybe no match implicitly produces exception
        }

This example doesn't make sense. You are matching type parameter itself and then use `mappingForm` which is indeed of that type. Actually, this method even doesn't need to be generic at all. You should perform pattern-matching on `mappingForm` itself. 
```cs
        private Target Map(object mappingFrom) => mappingFrom switch( 
            case Foo foo: new FooMapper().Map(foo),
            case Bar bar: new BarMapper().Map(bar),
            case *: throw new InvalidOprationException()
        );
```

**Comment by colin-jack** [comment]
I'm not sure I follow, you say the example doesn't make sense but your issues seems to just be with the specific syntax. I'll thus assume it is the syntax that's the issue...

I'm not sure I really follow what you mean about matching type parameter. On generics though, yup it doesn't need to be generic in that case. However at time I logged the issue there was no clear syntax for any of this and for the single example I chose to make the method generic just in case you wanted to apply additional constraints. I've not thought about that aspect much though.

Your example looks odd to my eyes, I understand it but compared to the Scala matching approach I find it a bit ugly. I haven't kept up with what's being suggested though, so if that switch approach is what's chosen then I'd learn to live with it.

**Comment by alrz** [comment]
@colin-jack I'm not proposing any new syntax, this is the current syntax mentioned in the [spec draft](https://github.com/dotnet/roslyn/blob/future/docs/features/patterns.md). So nothing with your syntax or mine. the problem is with this line `TFrom match`, [there're chances that you could perform pattern matching on `Type`](https://github.com/dotnet/roslyn/issues/206#issuecomment-144395536), but in your example you don't need to, because you want the final variable of the pattern type, e.g. `Foo` or `Bar`.  My example above written according to the currently proposed spec for pattern matching and it does solve the problem that you've mentioned in this thread.

**Comment by colin-jack** [comment]
@alrz
On generics, I was working forward from case I had where I had similar problem, I still think it might be useful to add generic constraints, and I think limiting what objects can be passed in would be clear, but who knows how useful it'd be.

Get you on syntax, I didn't really mean that you'd made up new syntax. I have looked at the spec a couple of times but my main reaction is to wish it had a few more examples (like yours) in it (examples FTW).

So in summary if your example is as-per spec then it seems like it'd meet my needs, hurrah.

**Comment by alrz** [comment]
@colin-jack you would be able to use generics and constraints with no doubt, but you should switch on the parameter itself rather than its type. with constraints I presume your patterns also need to be *pattern compatible* whit the resultant constrainted type.

**Comment by colin-jack** [comment]
@alrz 
Brilliant ta, sounds like it'll meet my needs perfectly.

**Comment by gafter** [comment]
@colin-jack Since the existing form of pattern-matching addresses your use case, we're not going to take this suggestion to address your use case with pattern-matching on the type parameter.
